api/v1beta1/externaldns_types.go

```diff
@@ -0,0 +1,603 @@
+	// +optional
+	OpenShiftRoute *ExternalDNSOpenShiftRouteOptions `json:"openshiftRouteOptions,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=OpenShiftRoute;Service;CRD
+type ExternalDNSSourceType string
+
+const (
+	SourceTypeRoute   ExternalDNSSourceType = "OpenShiftRoute"
+	SourceTypeService ExternalDNSSourceType = "Service"
+	SourceTypeCRD     ExternalDNSSourceType = "CRD"
+)
+
+// +kubebuilder:validation:Enum=Ignore;Allow
+type HostnameAnnotationPolicy string
+
+const (
+	HostnameAnnotationPolicyIgnore HostnameAnnotationPolicy = "Ignore"
+	HostnameAnnotationPolicyAllow  HostnameAnnotationPolicy = "Allow"
+)
+
+// ExternalDNSServiceSourceOptions describes options
+// specific to the ExternalDNS service source.
+type ExternalDNSServiceSourceOptions struct {
+	// ServiceType determines what types of Service resources
+	// are watched by ExternalDNS. The following types are
+	// available options:
+	//
+	//  "NodePort"
+	//  "ExternalName"
+	//  "LoadBalancer"
+	//  "ClusterIP"
+	//
+	// One or more Service types can be specified, if desired.
+	//
+	// Note that using the "ClusterIP" service type will enable
+	// the ExternalDNS "--publish-internal-services" flag,
+	// which allows ExternalDNS to publish DNS records
+	// for ClusterIP services.
+	//
+	// If no service types are provided, ExternalDNS will be
+	// configured to create DNS records for LoadBalancer services
+	// only by default.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:={"LoadBalancer"}
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	ServiceType []corev1.ServiceType `json:"serviceType,omitempty"`
+}
+
+type ExternalDNSOpenShiftRouteOptions struct {
+	// RouterName is the name of a router (AKA ingress controller) as
+	// reported in Route.status.ingress[].routerName.  External-dns will use
+	// the canonical hostname of the router identified by this name when
+	// publishing records for a given route.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	RouterName string `json:"routerName"`
+}
+
+// ExternalDNSCRDSourceOptions describes options for configuring
+// the ExternalDNS CRD source. The ExternalDNS CRD Source implementation
+// expects CRD resources to have specific fields, including a DNSName field. See
+// https://github.com/kubernetes-sigs/external-dns/blob/master/docs/contributing/crd-source.md
+// for more information.
+//
+// A configured CRD source would grant precise control to external DNS resources
+// to any user who can create/update/delete the given CRD.
+type ExternalDNSCRDSourceOptions struct {
+	// Kind is the kind of the CRD
+	// source resource type to be
+	// consumed by ExternalDNS.
+	//
+	// e.g. "DNSEndpoint"
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Kind string `json:"kind"`
+
+	// Version is the API version
+	// of the given resource kind for
+	// ExternalDNS to use.
+	//
+	// e.g. "externaldns.k8s.io/v1alpha1"
```

Comment:
```suggestion
	// e.g. "externaldns.k8s.io/v1beta1"
```

Reply:
It is the version of the [upstream CRD](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/contributing/crd-source/crd-manifest.yaml#L20), not our ExternalDNS API.